### ISSUE TITLE
[desktop] High memory consumption [Part 2]

### ISSUE THREAD

**Issue by ivk1800** [flutter/flutter#95092]
continue of https://github.com/flutter/flutter/issues/90868

in multi modules project consumes too much memory.  project contains 40 modules with included ```dart_code_metrics```. I attach test project for reproduce.

Steps fo reproduce:
1. download archive and extract
2. open project in intellij idea
3. open any dart file and enable dart for project
4. run sync.sh
5. close idea
6. open again project
7. wait until analizier done
8. check system monitor
9. check memory, memory grows
10. wait until stop growing
11. memory not released

```dart_code_metrics: 4.8.1```

Dart version 2.15.0, macOS 11.5.2 20G95 darwin-x64

![image](https://user-images.githubusercontent.com/11444783/145688871-4f5d3881-6821-430f-adbf-1979b36fcc62.png)

[HighMemoryConsumption.zip](https://github.com/flutter/flutter/files/7697599/HighMemoryConsumption.zip)

https://user-images.githubusercontent.com/11444783/145688588-5680b08d-3b1a-489a-9d93-c7ce92ab07cc.mov

**Comment by ivk1800** [comment]
@scheglov @incendial

**Comment by incendial** [comment]
@ivk1800 we have [an open issue](https://github.com/dart-code-checker/dart-code-metrics/issues/528) in Dart Code Metrics repo and right now it looks like the problem is with the plugin, not the integration. But I'd really appreciate some help from the Dart team, just to be sure that the integration is solid and doesn't have memory leaks.

**Comment by incendial** [comment]
After some investigation with different plugin setups I'm not that sure that it's our code leaking. Definitely need some help with checking the integration memory consumption.

**Comment by darshankawar** [comment]
I tried the given zip and followed steps to replicate on latest master and stable on mac and can see CPU and memory consumed by dart, as below:

![Screenshot 2021-12-13 at 2 19 43 PM](https://user-images.githubusercontent.com/67046386/145782138-ed61043b-2738-4ef0-9d5b-f5884c80c217.png)
![Screenshot 2021-12-13 at 2 19 58 PM](https://user-images.githubusercontent.com/67046386/145782146-1c4e7408-daed-4e5c-af56-e1b98d0a5e14.png)

<details>
<summary> stable, master flutter doctor -v </summary>

```
[✓] Flutter (Channel stable, 2.8.0, on Mac OS X 10.15.4 19E2269 darwin-x64,
    locale en-GB)
    • Flutter version 2.8.0 at /Users/dhs/documents/fluttersdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision cf44000065 (10 hours ago), 2021-12-08 14:06:50 -0800
    • Engine revision 40a99c5951
    • Dart version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 30)
    • Android SDK at /Users/dhs/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Users/dhs/Library/Android/sdk
    • Java binary at: /Users/dhs/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/202.7486908/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.    

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Users/dhs/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/202.7486908/Android
      Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)        

[✓] VS Code (version 1.57.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (4 available)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 93.0.4577.82


• No issues found!

[[✓] Flutter (Channel master, 2.6.0-12.0.pre.1002, on Mac OS X 10.15.4 19E2269
    darwin-x64, locale en-GB)
    • Flutter version 2.6.0-12.0.pre.1002 at
      /Users/dhs/documents/fluttersdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 9de8849905 (3 hours ago), 2021-12-12 20:54:15 -0500
    • Engine revision f16c96c31f
    • Dart version 2.16.0 (build 2.16.0-85.0.dev)
    • DevTools version 2.9.1
    
[✓] Android toolchain - develop for Android devices (Android SDK version 30)
    • Android SDK at /Users/dhs/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Users/dhs/Library/Android/sdk
    • Java binary at: /Users/dhs/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/202.7486908/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.    

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5.1, Build version 12E507
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Users/dhs/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/202.7486908/Android
      Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)        

[✓] VS Code (version 1.57.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[[✓] Connected device (4 available)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 93.0.4577.82


• No issues found!


```

</details>

**Comment by mraleph** [comment]
AFAIK plugins run in isolates so I would not be surprised that memory jumps considerably when you load additional analyser plugins. AST representation is likely duplicated between main analyser isolate and each plugin. 

@bwilkerson @scheglov should know how to diagnose this further.

**Comment by scheglov** [comment]
OK, I also see high memory consumption, by the plugins isolate.
![image](https://user-images.githubusercontent.com/384794/145854491-a22b415a-334b-4f3c-9f62-07c6c0c9d56a.png)

**Comment by scheglov** [comment]
Looking into individual `AnalysisDriver` instance shows that both have approximately the same size for `FileSystemState`, but plugins have much more. Specifically:
1. Each `AnalysisDriver` in plugins has its own `MemoryByteStore`, which means that each of them analysis Flutter independently. This wastes both CPU and memory. Ideally we want to use one shared `ByteStore` instance.
2. Each driver also has a large portion of the heap used by the element model. This might be something that I affected recently - we started to keep `LibraryContext` instances even after analysis is done, to support better code completions. But you don't need this in plugins. Or it could also be because of (1) - we have to resynthesize portions of the element model to link more element models, even though we could just resynthisize all of it from bytes without linking at all.

![image](https://user-images.githubusercontent.com/384794/145859692-925aa39d-a994-4919-9972-a4d108040d87.png)

Main:
![image](https://user-images.githubusercontent.com/384794/145859790-0c77d354-9b59-42de-8460-95e5c4810214.png)

**Comment by bwilkerson** [comment]
> ... to support better code completions. But you don't need this in plugins.

Not sure whether it matters, but plugins are allowed to also contribute to code completions.

**Comment by scheglov** [comment]
It probably does not matter for plugins. We have to keep element models to ensure that element instances are mostly the same, so we can use `Expando` to cache corresponding completion suggestions, which is important only when we suggest a lot - like a few thousand classes in Flutter. Building a few suggestions for `expr.^` context should not be very expensive anyway.

**Comment by bwilkerson** [comment]
> So, as he moves the caret in the file, we want to produce Quick Assists, and don't want to resolve file each time. It is not obvious to me that plugins need such caching.

Probably the most common use cases for plugins are to provide custom lints and custom fixes. It would be better if the plugin didn't need to re-resolve the priority file when the user clicks in a highlighted error range to get quick fixes.

**Comment by scheglov** [comment]
If the plugin does provide quick fixes - yes. But if only diagnostics - no.

Disabling caching priority files in plugins makes it somewhat better, but not totally. We still use a lot of heap, but most of it is garbage.

Before GC:
![image](https://user-images.githubusercontent.com/384794/145890129-55631960-b3e4-4cf8-b578-6f4da887c19f.png)

After GC:
![image](https://user-images.githubusercontent.com/384794/145890162-9313e7d5-ed32-4b47-bb32-4317503bfb9b.png)

It is unfortunate that VM continues to hold most of this heap.

**Comment by scheglov** [comment]
Using single static `MemoryByteStore`, shared between all drivers created in the plugin improves heap usage, and also makes it much, much faster (no number). Which is not surprising - we don't have to link large libraries of Flutter 40 times, and usually do it only once.

Before GC:
![image](https://user-images.githubusercontent.com/384794/145892210-56695976-a990-4e3c-8fb3-288bb0332cb4.png)

After GC:
![image](https://user-images.githubusercontent.com/384794/145892284-d8d37bf4-b7db-4fde-a591-86dafae60baa.png)

![image](https://user-images.githubusercontent.com/384794/145892380-e54a15f5-9455-42e3-9d10-b3cd34c7065a.png)

**Comment by incendial** [comment]
@scheglov first of all, thank you so much for the investigation!

> Using single static MemoryByteStore, shared between all drivers created in the plugin improves heap usage, and also makes it much, much faster (no number). Which is not surprising - we don't have to link large libraries of Flutter 40 times, and usually do it only once.

Could you please share how to achieve this with the current API. Right now we use `ContextBuilder(resourceProvider: resourceProvider);` [to instantiate the context](https://github.com/dart-code-checker/dart-code-metrics/blob/master/lib/src/analyzer_plugin/analyzer_plugin.dart#L75), which doesn't take anything else and [defined here](https://github.com/dart-lang/sdk/blob/main/pkg/analyzer/lib/dart/analysis/context_builder.dart#L18). Also, how to make MemoryByteStore sharable between all instances since the plugin is a separate dependency per package? Does they have some shared code too?

Another question - is it possible to share the `MemoryByteStore` with the main process?

> Disabling caching priority files in plugins makes it somewhat better, but not totally. We still use a lot of heap, but most of it is garbage.

I don't know, if it's related or not, but right now we need [to use hacks](https://github.com/dart-code-checker/dart-code-metrics/blob/master/lib/src/analyzer_plugin/analyzer_plugin.dart#L275) in order to highlight the errors we send to the analyzer in opened files. It was copied from another plugin. Maybe there is a better way to do it?

**Comment by m-skolnick** [comment]
@scheglov For what it's worth. I noticed the high memory consumption early this year. It's not something new to a recent release.

**Comment by scheglov** [comment]
I don't have great understanding of plugins and the APIs that they should be based on. I will leave it to @bwilkerson regarding plugins API. I just want to note that once you step inside the non-API territory with `DriverBasedAnalysisContext`, you could just as well started using `AnalysisContextCollectionImpl` that does allow you to set the `byteStore`.

As for which implementation of `ByteStore` to use, I used `MemoryByteStore` just for hack, but it is "infinite", will never flush, which you don't want to do long term. So, `MemoryCachingByteStore(NullByteStore)` is probably optimal. Using `EvictingFileByteStore` might let reusing some data between restarts, but might have its own complications.

Nothing of this is API of course.

My understanding of why you need these hacks is that when a file (that was `addFile` before) is affected by a `changeFile`, you want to get its resolved unit and produce additional diagnostics. This looks awfully like a lint rule. Which raises a question - why all this is not in `linter`?

If we stay in a plugin, a more direct way to achieve this is to a callback to `AnalysisDriver` that is invoked after a file resolution, e.g. `void Function(List<CompilationUnit>) onResolvedLibrary`. Then `results` will still contain only `ErrorsResult`s, but this is fine - you will not use it.

**Comment by incendial** [comment]
>  I just want to note that once you step inside the non-API territory with DriverBasedAnalysisContext, you could just as well started using AnalysisContextCollectionImpl that does allow you to set the byteStore

Unfortunately, we use implementation imports just because there is no other way to create a driver without them. We'd be really happy to use public API instead, but I got your point, I think we can at least give it a try in order to resolve current performance issue.

> As for which implementation of ByteStore to use, I used MemoryByteStore just for hack, but it is "infinite", will never flush, which you don't want to do long term. So, MemoryCachingByteStore(NullByteStore) is probably optimal. Using EvictingFileByteStore might let reusing some data between restarts, but might have its own complications.

Thank you, do we need to define it outside of `createAnalysisDriver` method?

> Nothing of this is API of course.

But why? Are you not interested in investing into plugins API or it's just so low on the priority list and you don't have ti,e for it? I mean, I understand that it's barely used right now, but the potential of custom lint rules for a specific team requirements is underestimated right now.

> Which raises a question - why all this is not in `linter`?

You mean, why the functionality that the package provides is not added to the linter package, instead of a separate package with custom lint rules?

**Comment by incendial** [comment]
@scheglov @bwilkerson sorry for pinging, but could you please address questions above?

**Comment by frankforpresident** [comment]
Any update on this issue?

**Comment by bwilkerson** [comment]
Sorry for the long delay. I suspect that Konstantin was expecting that I'd answer it, but between the holidays and an illness I've been out for a while.

> ... do we need to define it outside of `createAnalysisDriver` method?

If you use `AnalysisContextCollectionImpl` then you can pass in a byte store that will be shared across all of the contexts, which is what I suspect that Konstantin was thinking about.

When we designed the plugin API we didn't yet have any way for plugins to duplicate the logic that the server used to determine where the context roots are. In fact, the logic was, at that time, largely duplicated in both the analysis server and the `dartanalyzer`, but the two implementations were not completely consistent, which led to consistency issues between the tools.

We didn't want plugins to also be inconsistent, so we added the `analysis.setContextRoots` protocol rather than having plugins implement the `analysis.setAnalysisRoots` from the server's protocol.

Now that we have `AnalysisContextCollection` as a public API, and given that it's used by the analysis server, it could be used directly by plugins (though it would need to be enhanced a bit) and still maintain consistency (although version skew could still be an issue). I'm not sure what the best course of action is here, but we definitely need to have a simple and consistent way for plugins to create analysis contexts.

> > Nothing of this is API of course.
>
> But why? Are you not interested in investing into plugins API or it's just so low on the priority list and you don't have time for it? I mean, I understand that it's barely used right now, but the potential of custom lint rules for a specific team requirements is underestimated right now.

Personally, I agree with you. That's why I wrote the initial plugin support. :-)

But honestly, yes, it has been very low priority up to now. I believe that there's growing awareness on the Dart team that we really do need to support a plugin story of some kind, and we've started having some serious conversations about how we can address some of the memory usage and performance issues in order to make this a more viable feature.

I don't have any idea what the timeframe for this looks like, or even where these conversations will lead, but I also believe that there is an awareness that we need to have more of these discussions in forums where the community can help us understand better the requirements for such support. I'm hoping that the public discussions can start soon, but time will tell.

**Comment by incendial** [comment]
> If you use AnalysisContextCollectionImpl then you can pass in a byte store that will be shared across all of the contexts, which is what I suspect that Konstantin was thinking about.

So you mean that
```dart
final byteStore = _createByteStore();

class AnalyzerPlugin extends ServerPlugin {
  ...
 
  @override
  AnalysisDriverGeneric createAnalysisDriver(plugin.ContextRoot contextRoot) {
    ...
    // create AnalysisContextCollectionImpl with the shared byteStore
    ...
  }
}
```
will be enough for multiple plugins to share single instance of bytestore or we need something more?

> Personally, I agree with you. That's why I wrote the initial plugin support. :-)
>
> But honestly, yes, it has been very low priority up to now. I believe that there's growing awareness on the Dart team that we > really do need to support a plugin story of some kind, and we've started having some serious conversations about how we > can address some of the memory usage and performance issues in order to make this a more viable feature.
>
> I don't have any idea what the timeframe for this looks like, or even where these conversations will lead, but I also believe 
> that there is an awareness that we need to have more of these discussions in forums where the community can help us
> understand better the requirements for such support. I'm hoping that the public discussions can start soon, but time will tell.

Thank you for answering! I'm really happy to see awareness growing and looking forward to public discussions, we're ready to help from our experience with the current plugins api state. 

I think that in order to get more feedback from the community, the community needs to understand the real value of that api, but from my point of view it's barely used because it's not that easy to start with (no examples), lack of examples of real use-cases for day to day development and a lot of different issues down the way (like performance, private api usage, hack for priority files, etc.). So, I think that there won't be many sources of feedback, but I might be wrong.

**Comment by incendial** [comment]
@bwilkerson could you please take a look on the code example above?

**Comment by incendial** [comment]
@scheglov sorry for bothering, but what did you mean by `Using single static MemoryByteStore, shared between all drivers created`? I tried move the byteStore declaration out of the driver class and didn't see any memory improvements https://github.com/dart-code-checker/dart-code-metrics/pull/676/files#diff-f63386d01dd721283457671270dec6369293337bf7f3eb190565785d5b1b81bcR28.

Could you please take a look and tell, what's wrong?

**Comment by incendial** [comment]
@bwilkerson @scheglov any change you can reply?

**Comment by fzyzcjy** [comment]
Hi, is there any updates? I am seeing as large as 15GB of memory usage of the dart analyzer, which is almost 4x a big Intellij IDEA ide!

**Comment by incendial** [comment]
Any updates?

**Comment by frankforpresident** [comment]
Would love to see a fix for this issue, it is really hurting my productivity

**Comment by incendial** [comment]
> Using single static `MemoryByteStore`, shared between all drivers created in the plugin improves heap usage, and also makes it much, much faster (no number). Which is not surprising - we don't have to link large libraries of Flutter 40 times, and usually do it only once.
> 
> Before GC: ![image](https://user-images.githubusercontent.com/384794/145892210-56695976-a990-4e3c-8fb3-288bb0332cb4.png)
> 
> After GC: ![image](https://user-images.githubusercontent.com/384794/145892284-d8d37bf4-b7db-4fde-a591-86dafae60baa.png)
> 
> ![image](https://user-images.githubusercontent.com/384794/145892380-e54a15f5-9455-42e3-9d10-b3cd34c7065a.png)

The problem is that right now each plugin creates its own representation of all dependencies (like Flutter) in the memory.  Looks like there is a solution to that problem, but I don't understand what `Using single static `MemoryByteStore`` means and how exactly it should be implemented. I tried different approaches and non of them helped. And there is no public API to achieve that.

**Comment by leafpetersen** [comment]
So is it correct to summarize the current status as follows?
  - Code metrics uses the public plugin API (or does it also use some non-public API?  I'm confused about this)
  - We believe (for some value of we?) that that API does not provide the functionality that Code metrics needs to be performant.
    - Q: Is this just code metrics?  Or something general to plugins?
  - There may be (is?) some extension to the public API that could allow code metrics to perform better.
    - Q: How clear are we on what that API would look like?
    - Q: Do we think this API is a good thing, something we want to support?
    - Q: How much short term effort is it to implement that API?
    - Q: What's the long term maintenance burden we'd be taking on by adding that API?
    - Q: Is this something that we (the Dart team) would need to budget for, or is this something that external contributors (e.g. Code metrics?) could pick up?

I'm very concerned about getting pulled into a quagmire here, where each incremental step makes sense but either we end up in a place with either an API surface that isn't what we want, or worse we end up with an API that we don't have the resources to support (we've fallen into this trap before, and I'm keen to avoid it).

**Comment by incendial** [comment]
> Code metrics uses the public plugin API (or does it also use some non-public API? I'm confused about this)

We do use some private API too, coz there is just no way to create a plugin without it.

> We believe (for some value of we?) that that API does not provide the functionality that Code metrics needs to be performant.
Q: Is this just code metrics? Or something general to plugins?

As mentioned in the discussion, there is no way to pass ByteStore using current API. I've tried to pass it using private API, but nothing changed. I think it's general to plugins since it's about caching resolved dependencies.

I guess I can't answer other questions, but an alternative to changing the API would be discussing the exact changes that should be applied. Maybe we (as Dart Code Metrics team) can just implement them on our side using private API and other plugins in the future will be able to see our code and do the same. But I still don't understand what exactly should be changed in the code in order to make it work right.

**Comment by frankforpresident** [comment]
@leafpetersen @incendial 

What is the status of this issue at the moment?

**Comment by scheglov** [comment]
Here is a quickly hacked plugin, which I think should work with the current version of `analyzer` and `analyzer_plugin`.

But this is not how I would like authors to write plugins. In my view you should use `AnalysisContextCollection`, create it in your plugin, and we should remove `createAnalysisDriver` method altogether now. I did an experiment with a changed version of `ServerPlugin` and such `AnalysisContextCollectionImpl` based plugin, and it seems to work fine. It will take some time to make it production ready though.

<details><summary>Hacked plugin</summary>
<p>

```diff
Index: pubspec.yaml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/pubspec.yaml b/pubspec.yaml
--- a/pubspec.yaml	(revision 5260f2a54c9bbe11bcf3815e02c7eb46bf7b2a0b)
+++ b/pubspec.yaml	(date 1651259756522)
@@ -34,3 +34,11 @@
 
 executables:
   metrics:
+
+dependency_overrides:
+  _fe_analyzer_shared:
+    path: /Users/scheglov/Source/Dart/sdk.git/sdk/pkg/_fe_analyzer_shared
+  analyzer:
+    path: /Users/scheglov/Source/Dart/sdk.git/sdk/pkg/analyzer
+  analyzer_plugin:
+    path: /Users/scheglov/Source/Dart/sdk.git/sdk/pkg/analyzer_plugin
Index: lib/src/analyzer_plugin/analyzer_plugin.dart
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/lib/src/analyzer_plugin/analyzer_plugin.dart b/lib/src/analyzer_plugin/analyzer_plugin.dart
--- a/lib/src/analyzer_plugin/analyzer_plugin.dart	(revision 5260f2a54c9bbe11bcf3815e02c7eb46bf7b2a0b)
+++ b/lib/src/analyzer_plugin/analyzer_plugin.dart	(date 1651260365388)
@@ -1,6 +1,7 @@
 // ignore_for_file: public_member_api_docs
 
 import 'dart:async';
+import 'dart:io' as io;
 
 import 'package:analyzer/dart/analysis/context_locator.dart';
 import 'package:analyzer/dart/analysis/results.dart';
@@ -12,6 +13,7 @@
 import 'package:analyzer/src/dart/analysis/context_builder.dart';
 // ignore: implementation_imports
 import 'package:analyzer/src/dart/analysis/driver.dart';
+import 'package:analyzer/src/dart/analysis/file_content_cache.dart';
 import 'package:analyzer_plugin/plugin/plugin.dart';
 import 'package:analyzer_plugin/protocol/protocol_generated.dart' as plugin;
 
@@ -30,6 +32,8 @@
 class AnalyzerPlugin extends ServerPlugin {
   static const _analyzer = LintAnalyzer();
 
+  late final FileContentCache fileContentCache;
+
   final _configs = <AnalysisDriverGeneric, LintAnalysisConfig>{};
 
   var _filesFromSetPriorityFilesRequest = <String>[];
@@ -47,24 +51,44 @@
   @override
   String get version => '1.0.0-alpha.0';
 
-  AnalyzerPlugin(ResourceProvider provider) : super(provider);
+  AnalyzerPlugin(ResourceProvider provider) : super(provider) {
+    fileContentCache = FileContentCache(resourceProvider);
+  }
 
   @override
   void contentChanged(String path) {
-    super.driverForPath(path)?.addFile(path);
+    // super.driverForPath(path)?.addFile(path);
+    logLine('[contentChanged][path: $path]');
+    final driver = super.driverForPath(path);
+    if (driver is AnalysisDriver) {
+      driver.changeFile(path);
+      runZonedGuarded(
+        () async {
+          final affectedFiles = await driver.applyPendingFileChanges();
+          await _analyzeFiles(driver, affectedFiles);
+        },
+        (e, stackTrace) {
+          logLine('[exception][e: $e][stackTrace: $stackTrace]');
+          channel.sendNotification(
+            plugin.PluginErrorParams(false, e.toString(), stackTrace.toString())
+                .toNotification(),
+          );
+        },
+      );
+    }
   }
 
   @override
   AnalysisDriverGeneric createAnalysisDriver(plugin.ContextRoot contextRoot) {
     final rootPath = contextRoot.root;
-    final locator =
+    final contextRoots =
         ContextLocator(resourceProvider: resourceProvider).locateRoots(
       includedPaths: [rootPath],
       excludedPaths: contextRoot.exclude,
       optionsFile: contextRoot.optionsFile,
     );
 
-    if (locator.isEmpty) {
+    if (contextRoots.isEmpty) {
       final error = StateError('Unexpected empty context');
       channel.sendNotification(plugin.PluginErrorParams(
         true,
@@ -75,10 +99,14 @@
       throw error;
     }
 
+    final contextRoot2 = contextRoots.first;
+    logLine('[create][contextRoots: ${contextRoots.map((e) => e.root)}]');
+
     final builder = ContextBuilderImpl(resourceProvider: resourceProvider);
     final context = builder.createContext(
-      contextRoot: locator.first,
+      contextRoot: contextRoot2,
       byteStore: _byteStore,
+      fileContentCache: fileContentCache,
     );
     final dartDriver = context.driver;
     final config = _createConfig(dartDriver, rootPath);
@@ -102,14 +130,17 @@
     // }
 
     runZonedGuarded(
-      () {
-        dartDriver.results.listen((analysisResult) {
-          if (analysisResult is ResolvedUnitResult) {
-            _processResult(dartDriver, analysisResult);
-          }
-        });
+      () async {
+        final filesToAnalyze = contextRoot2.analyzedFiles().toList();
+        await _analyzeFiles(dartDriver, filesToAnalyze);
+        // dartDriver.results.listen((analysisResult) {
+        //   if (analysisResult is ResolvedUnitResult) {
+        //     _processResult(dartDriver, analysisResult);
+        //   }
+        // });
       },
       (e, stackTrace) {
+        logLine('[exception][e: $e][stackTrace: $stackTrace]');
         channel.sendNotification(
           plugin.PluginErrorParams(false, e.toString(), stackTrace.toString())
               .toNotification(),
@@ -120,6 +151,37 @@
     return dartDriver;
   }
 
+  Future<void> _analyzeFiles(
+    AnalysisDriver dartDriver,
+    List<String> filesToAnalyze,
+  ) async {
+    logLine(
+      '[analyzeFiles][root: ${dartDriver.name}]'
+      '[filesToAnalyze: $filesToAnalyze]',
+    );
+    final analysisSession = dartDriver.currentSession;
+    for (final path in filesToAnalyze) {
+      if (path.endsWith('.dart')) {
+        logLine('  [path: $path]');
+        final analysisResult = await analysisSession.getResolvedUnit(path);
+        if (analysisResult is ResolvedUnitResult) {
+          logLine('    [has ResolvedUnitResult]');
+          _processResult(dartDriver, analysisResult);
+          logLine('    processed');
+        }
+      }
+    }
+    dartDriver.clearLibraryContext();
+  }
+
+  void logLine(String line) {
+    io.File('/Users/scheglov/tmp/plugin_log.txt').writeAsStringSync(
+      '$line\n',
+      mode: io.FileMode.append,
+      flush: true,
+    );
+  }
+
   @override
   Future<plugin.AnalysisSetContextRootsResult> handleAnalysisSetContextRoots(
     plugin.AnalysisSetContextRootsParams parameters,
@@ -278,29 +340,29 @@
   /// As a result, [_processResult] will get called with resolved units, and thus all of our diagnostics
   /// will get run on all files in the repo instead of only the currently open/edited ones!
   void _updatePriorityFiles() {
-    final filesToFullyResolve = {
-      // Ensure these go first, since they're actually considered priority; ...
-      ..._filesFromSetPriorityFilesRequest,
-
-      // ... all other files need to be analyzed, but don't trump priority
-      for (final driver2 in driverMap.values)
-        ...(driver2 as AnalysisDriver).addedFiles,
-    };
-
-    // From ServerPlugin.handleAnalysisSetPriorityFiles.
-    final filesByDriver = <AnalysisDriverGeneric, List<String>>{};
-    for (final file in filesToFullyResolve) {
-      final contextRoot = contextRootContaining(file);
-      if (contextRoot != null) {
-        // TODO(dkrutskikh): Which driver should we use if there is no context root?
-        final driver = driverMap[contextRoot];
-        if (driver != null) {
-          filesByDriver.putIfAbsent(driver, () => <String>[]).add(file);
-        }
-      }
-    }
-    filesByDriver.forEach((driver, files) {
-      driver.priorityFiles = files;
-    });
+    // final filesToFullyResolve = {
+    //   // Ensure these go first, since they're actually considered priority; ...
+    //   ..._filesFromSetPriorityFilesRequest,
+    //
+    //   // ... all other files need to be analyzed, but don't trump priority
+    //   for (final driver2 in driverMap.values)
+    //     ...(driver2 as AnalysisDriver).addedFiles,
+    // };
+    //
+    // // From ServerPlugin.handleAnalysisSetPriorityFiles.
+    // final filesByDriver = <AnalysisDriverGeneric, List<String>>{};
+    // for (final file in filesToFullyResolve) {
+    //   final contextRoot = contextRootContaining(file);
+    //   if (contextRoot != null) {
+    //     // TODO(dkrutskikh): Which driver should we use if there is no context root?
+    //     final driver = driverMap[contextRoot];
+    //     if (driver != null) {
+    //       filesByDriver.putIfAbsent(driver, () => <String>[]).add(file);
+    //     }
+    //   }
+    // }
+    // filesByDriver.forEach((driver, files) {
+    //   driver.priorityFiles = files;
+    // });
   }
 }
Index: tools/analyzer_plugin/pubspec.yaml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/tools/analyzer_plugin/pubspec.yaml b/tools/analyzer_plugin/pubspec.yaml
--- a/tools/analyzer_plugin/pubspec.yaml	(revision 5260f2a54c9bbe11bcf3815e02c7eb46bf7b2a0b)
+++ b/tools/analyzer_plugin/pubspec.yaml	(date 1651259756525)
@@ -10,3 +10,13 @@
 
 dev_dependencies:
   lints: ^1.0.1
+
+dependency_overrides:
+  dart_code_metrics:
+    path: /Users/scheglov/dart/dart-code-metrics
+  _fe_analyzer_shared:
+    path: /Users/scheglov/Source/Dart/sdk.git/sdk/pkg/_fe_analyzer_shared
+  analyzer:
+    path: /Users/scheglov/Source/Dart/sdk.git/sdk/pkg/analyzer
+  analyzer_plugin:
+    path: /Users/scheglov/Source/Dart/sdk.git/sdk/pkg/analyzer_plugin
```

</p>
</details>

Heap usage (still on 10 out of 40 target packages):
<img width="1334" alt="image" src="https://user-images.githubusercontent.com/384794/166059240-a9abff3f-1b5f-4015-9db4-d88330ef3804.png">

This improvement because of two changes:
1. Using shared `FileContentCache` instance, so that we have only one copy of a file content (think about many Flutter package files).
2. We call `dartDriver.clearLibraryContext()`, so drivers don't keep element models. This reduces heap usage, but at cost of being slower during a next analysis. Theoretically it is up to you when to call it, I can imagine a resettable timer for 30 seconds, so that when the user actively works in a package, we keep its elements hot and ready, but once there is a pause, we compact. Nobody writes 40 packages at once (I think :-)), so at worst we will keep 1 or 2 packages.

**Comment by ivk1800** [comment]
Dart SDK version: 2.16.2 (stable) (Tue Mar 22 13:15:13 2022 +0100) on "macos_x64"
dart_code_metrics: 4.14.0

I repeat steps for reproduce.
Now dart process keep 4-5gb insteaв 12gb

**Comment by scheglov** [comment]
I believe the improvement you see comes from using shared `ByteStore` in the recent versions of `dart_code_metrics`.

**Comment by jacob314** [comment]
> Dart SDK version: 2.16.2 (stable) (Tue Mar 22 13:15:13 2022 +0100) on "macos_x64" dart_code_metrics: 4.14.0
> 
> I repeat steps for reproduce. Now dart process keep 4-5gb instead 12gb

Glad to hear the memory usage went down significantly.
For the case where your Dart process uses 4-5gb, how much of the memory is your analysis plugin using and how much is the core analyzer? Based on #s from when @scheglov ran your plugin locally, with his optimizations, your plugin is likely now using half the memory usage of the core analyzer which is about as memory efficient as it is reasonable to expect a plugin to be without a major rewrite to how plugins are supported. The plugin would use the same as the core analyzer if it wasn't for the optimizations @scheglov added to flush the element models more aggressively in the plugin than is possible when supporting completion and quick fixes as in the core plugin.

Screenshot from @scheglov showing how memory usage is split between the plugin and core analyzer:
<img width="1648" alt="image (2)" src="https://user-images.githubusercontent.com/1226812/166073234-4e6f64be-5301-4e34-9281-804d1b3d3248.png">

**Comment by incendial** [comment]
> I don't quite understand what is the issue we are trying to solve here.
>
> High memory consumption as mentioned in this issue title?
> 
> I see that you already switched to using shared ByteStore and provide it to ContextBuilderImpl.createContext(). This helps  with performance - when I patch the plugin to not provide it, it takes many times longer to analyze the target set of packages.
>
> But you say that nothing changed, so either you or me measure CPU time incorrectly, or performance is not the issue that you worry about in this issue the most.
> 
> The implementation details, like using or not internal APIs, will follow from the goal we want to achieve.
>
> And, looking into heap allocation, I see that heap usage improved too.
> But I guess this is not enough?

The problem is that even after changing the `ByteStore` usage (it became available for our users like ~3 month ago), we still received several reports about memory problems and our attempts to measure it properly also shown that nothing significantly changed (guess we just measured it wrong 😞 ). I'll ask our users again (but since @ivk1800 mentioned the improvement, hope they'll also confirm it).

@fzyzcjy may I ask you to also recheck, if this problem is still bothering you?

@scheglov thank you very much!

> The plugin would use the same as the core analyzer if it wasn't for the optimizations @scheglov added to flush the element models more aggressively in the plugin than is possible when supporting completion and quick fixes as in the core plugin.

Sorry, but I didn't quite get the idea, what changes are mentioned there? Are you talking about new API `applyPendingFileChanges` or there are some other changes? If there is some other changes, from which Dart SDK version they are available?

**Comment by incendial** [comment]
> Here is a quickly hacked plugin, which I think should work with the current version of `analyzer` and `analyzer_plugin`.
> 
> But this is not how I would like authors to write plugins. In my view you should use `AnalysisContextCollection`, create it in your plugin, and we should remove `createAnalysisDriver` method altogether now. I did an experiment with a changed version of `ServerPlugin` and such `AnalysisContextCollectionImpl` based plugin, and it seems to work fine. It will take some time to make it production ready though.
> 
> Hacked plugin
> Heap usage (still on 10 out of 40 target packages): <img alt="image" width="1334" src="https://user-images.githubusercontent.com/384794/166059240-a9abff3f-1b5f-4015-9db4-d88330ef3804.png">
> 
> This improvement because of two changes:
> 
> 1. Using shared `FileContentCache` instance, so that we have only one copy of a file content (think about many Flutter package files).
> 2. We call `dartDriver.clearLibraryContext()`, so drivers don't keep element models. This reduces heap usage, but at cost of being slower during a next analysis. Theoretically it is up to you when to call it, I can imagine a resettable timer for 30 seconds, so that when the user actively works in a package, we keep its elements hot and ready, but once there is a pause, we compact. Nobody writes 40 packages at once (I think :-)), so at worst we will keep 1 or 2 packages.
> 3. I removed the hack for priority files. As I described earlier, this forces the analyzer to keep resolved units, so all element models, in a `Map`. With this `clearLibraryContext()` would not help of course. Instead we use newish method from `AnalysisDriver` (and `AnalysisContext`) - `applyPendingFileChanges()` which returns us files to analyze, and we _pull_ resolved units, instead of makes them to be pushed into us via `results`.

This look soooo lit 🔥 so happy to see those changes, thank you for sharing. 

By `It will take some time to make it production ready though.` you mean that the api is not ready from your side or that we need to properly adapt it? Asking, coz in the code sample you shared there are also overrides for dependencies and it looks like we can use them like that.

**Comment by incendial** [comment]
And if the new API is ready to be used, we'd love to use it even if it's "hacky" right now, just to improve our users overall experience.

**Comment by scheglov** [comment]
My current work in progress is here https://dart-review.googlesource.com/c/sdk/+/243161

**Comment by scheglov** [comment]
I made a few more changes to the CL, so that `MyServerPlugin` is quite easy now, and has just one method to implement analysis. It might help us if you try to check it out, make updates to see performance numbers and logs that fit you, and whether (when using dependency overrides) this version behaves better in terms of heap.

**Comment by incendial** [comment]
> I made a few more changes to the CL, so that `MyServerPlugin` is quite easy now, and has just one method to implement analysis. It might help us if you try to check it out, make updates to see performance numbers and logs that fit you, and whether (when using dependency overrides) this version behaves better in terms of heap.

Thank you! Sure, we'll try to migrate to this api and check the difference

**Comment by incendial** [comment]
@scheglov sorry for the ping, but maybe you have any details about when the new api will be released?

**Comment by incendial** [comment]
> Ah, actually I thought that you will able to use dependency override to try it before the release. But apparently you were waiting for us to publish it officially, which makes sense, but this is not what we expected.
> 
> @bwilkerson maybe we should actually publish.

@dkrutskikh mentioned, that he tested the api locally and everything was fine. I thought that you've discussed it already. IMO the next step is to use the published version in order to receive some real world feedback from our users. What do you think?

**Comment by scheglov** [comment]
I cleaned up [the CL](https://dart-review.googlesource.com/c/sdk/+/243161), so might be able to publish it.

**Comment by scheglov** [comment]
```
Successfully uploaded https://pub.dev/packages/analyzer_plugin "0.11.0".
```

**Comment by incendial** [comment]
@scheglov thank you for publishing the API! While working on the migration https://github.com/dart-code-checker/dart-code-metrics/pull/900 I faced two questions:
1. We have a concept of a config which previously was created per each driver. The config might change only when the user update analysis_options.yaml file which happens pretty rarely, so it'd great to have an ability to create it once for each context. Looks like the best place to create it now is `afterNewContextCollection`. How regularly the context collection is expected to be recreated? Is it efficient to create configs for each context in `afterNewContextCollection`?
2. Overriding `handleEditGetFixes` to provide custom fixes only sends a path to the file and the offset. In order to calculate the fixes I ended up saving `AnalysisContextCollection` to a class field to then get the context for the path, provided by `handleEditGetFixes` parameters. Is it the expected approach or I'm missing something?

Overall I removed A LOT of code which really makes me happy, thank you again for implementing the new API 🔥

**Comment by scheglov** [comment]
1. Yes, `AnalysisContextCollection` is created rarely, usually during `handleAnalysisSetContextRoots()`, i.e. when the user opens or closes directories in the IDE. So, it is fine to do some work where, even some heavy work. You probably can associate a config with a `AnalysisContext` using `Expando`.
2. I think the idea was that you don't need to discover `AnalysisContext` from the collection by its path, instead you call `getResolvedUnitResult`, and either get the resolved unit, or an exception. But I imagine that there might be a need to discover, if by some reason you don't need a resolved unit, so if you have such scenario that is not, let me know. Specifically in `handleEditGetFixes` you use the resolved unit.

**Comment by incendial** [comment]
@scheglov sorry for the delay, we've finally merged the migration and unfortunately there is one issue that looks kinda critical - some time ago we reported flickering on input for the plugin analysis issues and it's now back.

https://user-images.githubusercontent.com/21116742/179055214-1f851604-0df9-4d59-ac62-ce0fbcbf2e1a.mov

This only happens in VS Code. @DanTup hi, maybe you can take a look and tell, if it's the VS Code integration problem?

Another interesting part is that after some time (like ~2 mins or something) the plugin is being killed
<img width="950" alt="Screenshot 2022-07-14 at 22 16 38" src="https://user-images.githubusercontent.com/21116742/179055707-a1329971-8cf2-4fe8-ae91-c433e0f69269.png">

Clearing the ~/.dartServer/.plugin_manager didn't help. And I also think that we reported something similar before, but I don't remember what was the reason.

We merged the API support into the master branch https://github.com/dart-code-checker/dart-code-metrics or you can install the https://pub.dev/packages/dart_code_metrics/versions/4.17.0-dev.1 prerelease version.

**Comment by incendial** [comment]
@DanTup that's pretty interesting, I tried on the DCM repo itself, master branch. Could you please try on it?

**Comment by incendial** [comment]
@DanTup also, you were testing on the DCM trailing comma rule, right?

**Comment by DanTup** [comment]
@incendial yep, I was testing with the trailing comma. I can repro in dart-code-metrics repo. I enabled the server instrumentation log and it _seems_ like the plugin is sending an empty set immediately after:

```
/// IDE sends change to server
1658162334643:Req:{"jsonrpc"::"2.0","method"::"textDocument/didChange","params"::{"textDocument"::{"uri"::"file::///Users/danny/Dev/TestStuff/dart-code-metrics/lib/src/analyzer_plugin/analyzer_plugin.dart","version"::10},"contentChanges"::[{"range"::{"start"::{"line"::39,"character"::46},"end"::{"line"::39,"character"::47}},"rangeLength"::1,"text"::""}]},"clientRequestTime"::1658162334643}

// Server sends change to plugins
1658162334644:PluginReq:{"id"::"77","method"::"analysis.updateContent","params"::{"files"::{"/Users/danny/Dev/TestStuff/dart-code-metrics/lib/src/analyzer_plugin/analyzer_plugin.dart"::{"type"::"change","edits"::[{"offset"::1456,"length"::1,"replacement"::""}]}}}}:file::///Users/danny/.dartServer/.plugin_manager/fa3ca12664cc4af4a944320f1b8acd36/analyzer_plugin/bin/plugin.dart::

// Server sends diagnostics from its own analysis
1658162334688:Res:{"method"::"textDocument/publishDiagnostics","params"::{"uri"::"file::///Users/danny/Dev/TestStuff/dart-code-metrics/lib/src/analyzer_plugin/analyzer_plugin.dart","diagnostics"::[]},"jsonrpc"::"2.0"}

// Plugin generates diagnostics
1658162334690:PluginNoti:{"event"::"analysis.errors","params"::{"file"::"/Users/danny/Dev/TestStuff/dart-code-metrics/lib/src/analyzer_plugin/analyzer_plugin.dart","errors"::[{"severity"::"WARNING","type"::"LINT","location"::{"file"::"/Users/danny/Dev/TestStuff/dart-code-metrics/lib/src/analyzer_plugin/analyzer_plugin.dart","offset"::1414,"length"::42,"startLine"::40,"startColumn"::5,"endLine"::40,"endColumn"::47},"message"::"Prefer trailing comma.","correction"::"required ResourceProvider resourceProvider,","code"::"prefer-trailing-comma","url"::"https:://dartcodemetrics.dev/docs/rules/common/prefer-trailing-comma","hasFix"::true}]}}:file::///Users/danny/.dartServer/.plugin_manager/fa3ca12664cc4af4a944320f1b8acd36/analyzer_plugin/bin/plugin.dart::

// Server passes plugin diagnostics to client
1658162334690:Res:{"method"::"textDocument/publishDiagnostics","params"::{"uri"::"file::///Users/danny/Dev/TestStuff/dart-code-metrics/lib/src/analyzer_plugin/analyzer_plugin.dart","diagnostics"::[{"range"::{"start"::{"line"::39,"character"::4},"end"::{"line"::39,"character"::46}},"severity"::2,"code"::"prefer-trailing-comma","codeDescription"::{"href"::"https:://dartcodemetrics.dev/docs/rules/common/prefer-trailing-comma"},"source"::"dart","message"::"Prefer trailing comma.\nrequired ResourceProvider resourceProvider,"}]},"jsonrpc"::"2.0"}

// Plugin generates empty diagnostics
1658162334694:PluginNoti:{"event"::"analysis.errors","params"::{"file"::"/Users/danny/Dev/TestStuff/dart-code-metrics/lib/src/analyzer_plugin/analyzer_plugin.dart","errors"::[]}}:file::///Users/danny/.dartServer/.plugin_manager/fa3ca12664cc4af4a944320f1b8acd36/analyzer_plugin/bin/plugin.dart::

// Server passes empty list to client
1658162334694:Res:{"method"::"textDocument/publishDiagnostics","params"::{"uri"::"file::///Users/danny/Dev/TestStuff/dart-code-metrics/lib/src/analyzer_plugin/analyzer_plugin.dart","diagnostics"::[]},"jsonrpc"::"2.0"}
```

Could you try adding some logging to where you send diagnostics, to see if it is occurring in the plugin? If so, I would try logging up the tree to figure out a) why they're being sent twice and b) why the second set is empty. I don't know if it's related, but the log above is filtered to only lines with that path in, but there was generation of diagnostics for some other files in between the two diagnostic responses (the first with the diagnostics, and the second that was empty):

```
1658162334690:PluginNoti:{"event"::"analysis.errors","params"::{"file"::"/Users/danny/Dev/TestStuff/dart-code-metrics/lib/src/analyzer_plugin/analyzer_plugin.dart","errors"::[{"severity"::"WARNING","type"::"LINT","location"::{"file"::"/Users/danny/Dev/TestStuff/dart-code-metrics/lib/src/analyzer_plugin/analyzer_plugin.dart","offset"::1414,"length"::42,"startLine"::40,"startColumn"::5,"endLine"::40,"endColumn"::47},"message"::"Prefer trailing comma.","correction"::"required ResourceProvider resourceProvider,","code"::"prefer-trailing-comma","url"::"https:://dartcodemetrics.dev/docs/rules/common/prefer-trailing-comma","hasFix"::true}]}}:file::///Users/danny/.dartServer/.plugin_manager/fa3ca12664cc4af4a944320f1b8acd36/analyzer_plugin/bin/plugin.dart::
1658162334690:Res:{"method"::"textDocument/publishDiagnostics","params"::{"uri"::"file::///Users/danny/Dev/TestStuff/dart-code-metrics/lib/src/analyzer_plugin/analyzer_plugin.dart","diagnostics"::[{"range"::{"start"::{"line"::39,"character"::4},"end"::{"line"::39,"character"::46}},"severity"::2,"code"::"prefer-trailing-comma","codeDescription"::{"href"::"https:://dartcodemetrics.dev/docs/rules/common/prefer-trailing-comma"},"source"::"dart","message"::"Prefer trailing comma.\nrequired ResourceProvider resourceProvider,"}]},"jsonrpc"::"2.0"}
1658162334690:PluginNoti:{"event"::"analysis.errors","params"::{"file"::"/Users/danny/Dev/TestStuff/dart-code-metrics/lib/src/analyzer_plugin/analyzer_plugin_starter.dart","errors"::[]}}:file::///Users/danny/.dartServer/.plugin_manager/fa3ca12664cc4af4a944320f1b8acd36/analyzer_plugin/bin/plugin.dart::
1658162334690:Res:{"method"::"textDocument/publishDiagnostics","params"::{"uri"::"file::///Users/danny/Dev/TestStuff/dart-code-metrics/lib/src/analyzer_plugin/analyzer_plugin_starter.dart","diagnostics"::[]},"jsonrpc"::"2.0"}
1658162334690:PluginNoti:{"event"::"analysis.errors","params"::{"file"::"/Users/danny/Dev/TestStuff/dart-code-metrics/lib/analyzer_plugin.dart","errors"::[]}}:file::///Users/danny/.dartServer/.plugin_manager/fa3ca12664cc4af4a944320f1b8acd36/analyzer_plugin/bin/plugin.dart::
1658162334690:Res:{"method"::"textDocument/publishDiagnostics","params"::{"uri"::"file::///Users/danny/Dev/TestStuff/dart-code-metrics/lib/analyzer_plugin.dart","diagnostics"::[]},"jsonrpc"::"2.0"}
1658162334692:Res:{"method"::"textDocument/publishDiagnostics","params"::{"uri"::"file::///Users/danny/Dev/TestStuff/dart-code-metrics/lib/src/analyzer_plugin/analyzer_plugin_starter.dart","diagnostics"::[]},"jsonrpc"::"2.0"}
1658162334694:PluginNoti:{"event"::"analysis.errors","params"::{"file"::"/Users/danny/Dev/TestStuff/dart-code-metrics/lib/src/analyzer_plugin/analyzer_plugin.dart","errors"::[]}}:file::///Users/danny/.dartServer/.plugin_manager/fa3ca12664cc4af4a944320f1b8acd36/analyzer_plugin/bin/plugin.dart::
1658162334694:Res:{"method"::"textDocument/publishDiagnostics","params"::{"uri"::"file::///Users/danny/Dev/TestStuff/dart-code-metrics/lib/src/analyzer_plugin/analyzer_plugin.dart","diagnostics"::[]},"jsonrpc"::"2.0"}
```

**Comment by incendial** [comment]
Yeah, that looks strange. I have no clue, why diagnostics are sent twice, but looks like there only two ways, why we can respond in an empty array of issues for the same unit:
1. if the returned unit is not somehow a ResolvedUnitResult (but it should be) https://github.com/dart-code-checker/dart-code-metrics/blob/7e37e37629e6d0c5a97bc6e4af1d554cad711e8f/lib/src/analyzer_plugin/analyzer_plugin.dart#L78 <img width="581" alt="Screenshot 2022-07-18 at 21 01 33" src="https://user-images.githubusercontent.com/21116742/179564199-821afe67-4587-454b-9c97-337bc4621d7b.png">

2. if the config for the given analysis context root path is absent (which is highly unlikely since the first response returns an issue) https://github.com/dart-code-checker/dart-code-metrics/blob/7e37e37629e6d0c5a97bc6e4af1d554cad711e8f/lib/src/analyzer_plugin/analyzer_plugin.dart#L131
 
> Could you try adding some logging to where you send diagnostics, to see if it is occurring in the plugin?

Are you talking about some specific logging? Coz we're not very good at debugging the plugin and I can only suggest publishing the modified version to check for any difference.

**Comment by incendial** [comment]
Funny, but creating files worked 😄 . Never tried this before. Okay, I see the problem:

**path from analyzeFile:** /Users/dimannich/Desktop/code/master-dart-code-metrics/lib/src/analyzer_plugin/analyzer_plugin.dart
**context root path:** /Users/dimannich/Desktop/code/master-dart-code-metrics
**amount of issues:** 4

**path from analyzeFile:** /Users/dimannich/Desktop/code/master-dart-code-metrics/lib/src/analyzer_plugin/analyzer_plugin.dart
**context root path:** /Users/dimannich/Desktop/code/master-dart-code-metrics/tools/analyzer_plugin
**amount of issues:** 0

So, it seems to that it's like one file is being analyzed by 2 context and the second one overrides the result of the first analysis. I'd say that behaviour is unexpected and looks the analysis context from tools/analyzer_plugin should not be used to analyze files not in the tools/analyzer_plugin.

**Comment by incendial** [comment]
And I still get the plugin removed after some time... 😞

**Comment by scheglov** [comment]
Could be https://github.com/dart-lang/sdk/issues/49404

**Comment by incendial** [comment]
> Could be [dart-lang/sdk#49404](https://github.com/dart-lang/sdk/issues/49404)

Actually, yeah, looks like it, thanks! Any ideas why the plugin might be removed after few minutes? We also faced a problem with the files not being analysed when they are not open, but we're collecting more info in that.

**Comment by incendial** [comment]
@DanTup thank you for the help!

**Comment by incendial** [comment]
After giving the new api to our users we received a report that the problem reported here: https://github.com/flutter/flutter/issues/90868 (with multiple plugin instances after `pub get`) has returned 😞:

```
1 run: 14.35 GB
2 run: 15.52 GB
3 run: 23.24 GB
4 run: 27.40 GB
5 run: 31.77 GB
```

@scheglov any chance you can take a look?

**Comment by incendial** [comment]
> After giving the new api to our users

Also received reports about flickering, now the plugin is not removed, but flickers or after some time just shows no errors.

Another problem that people started to report the issues from the newer package version when they haven't yet updated. Like they have DCM 4.17.0, but IDE somehow reports an issue from 4.18.0. I'm not sure, if it's even possible, but it's already two reports.

**Comment by scheglov** [comment]
It looks that we have a slightly different problem now.

As you might see below (I had to make it small), there are `EvictingFileByteStore._cacheCleanUpFunction` isolates stuck running. The reason for this is that `EvictingFileByteStore` is not supposed to be used by clients, IIRC it has no way to exit.

<img width="1705" alt="image" src="https://user-images.githubusercontent.com/384794/190009078-cea91799-5dc9-4bd4-9492-a3f21408a8da.png">

**Comment by incendial** [comment]
> It looks that we have a slightly different problem now.
> 
> As you might see below (I had to make it small), there are `EvictingFileByteStore._cacheCleanUpFunction` isolates stuck running. The reason for this is that `EvictingFileByteStore` is not supposed to be used by clients, IIRC it has no way to exit.
> 
> <img alt="image" width="1705" src="https://user-images.githubusercontent.com/384794/190009078-cea91799-5dc9-4bd4-9492-a3f21408a8da.png">

Oh... Do we need to call the cleanup from the plugin somewhere? We also use `EvictingFileByteStore` for our commands, do we have the same problem there?

Thank you for looking into that

**Comment by JELaVallee** [comment]
> Oh... Do we need to call the cleanup from the plugin somewhere? We also use `EvictingFileByteStore` for our commands, do we have the same problem there?

@incendial just ran into this same issue in a private custom-lint plugin that was also built with `EvictingFileByteStore` and when it was scanning a project with >40 root packages.

What we were observing: When `dart language-server` is running either in VSCode or Intellij/Android Studio, everytime we ran `flutter pub get` we would see ~5GB memory loss. Could easily get to ~30GB mem utilization just switching branches and refreshing pub a couple of times.

We fixed it by changing our `MemoryCachingByteStore` initialization to use `FileByteStore` as @scheglov noted:

```dart
    return MemoryCachingByteStore(
      FileByteStore(
        stateLocation.path,
        tempNameSuffix: DateTime.now().millisecondsSinceEpoch.toString(),
      ),
      memoryCacheSize,
    );
```

Now what we're seeing is that initial analysis after opening the IDE still adds about 4-5GB to the memory utilization of the `dart language-server` proc, but we can see it properly GC'ing the LRU cache in `MemoryCachingByteStore`s when they reach their `memoryCacheSize` max. 😄

**Comment by a-siva** [comment]
//cc @srawlins

**Comment by srawlins** [comment]
I believe we  can close this. This thread does not represent a single P1 issue. There is not a clear exit state. Much of the changes have been made to one Dart analyzer plugin. We've mitigated issues that arise from _multiple_ Dart analyzer plugins by limiting to one. We're working on a new plugin API. I just don't think there is much to track here.
